### ISSUE TITLE
incorrect inferred return for metaclass __getitem__

### ISSUE THREAD

**Issue by Levitanus** [PyCQA/pylint#2722]
In the code example below metaclass is used additionally for providing syntax sugar for instance-checking.
It is returning ArrType only within tuple passed to `__getitem__`.
MyPy also complains current code, but it can be fixed with @overload decorated function.
I can't find the way tell Pylint when and what class is returned here.
Also I can't understand, why at the same time it not complains on size attribute missing in ArrInt
```Python
import typing as ty

class ArrMeta(type):
    """Var getitem helper metaclass."""
    def __getitem__(cls,
                    ref: ty.Union[ty.Type[int], ty.Tuple[ty.Type[int], int]]
                    ) -> ty.Union[ty.Type['ArrInt'], ty.Type['ArrType']]:
        """Return concrete typed class of Arr."""
        size: ty.Optional[int] = None
        if isinstance(ref, tuple):
            size = ref[1]
            ref = ref[0]
            if size is not None:
                ArrType.size = size
                ArrType.ref_type = ref
                return ArrType
        if issubclass(ref, int):
            return ArrInt
        if issubclass(ref, float):
            return NotImplemented
        if issubclass(ref, str):
            return NotImplemented
        raise TypeError('subscribe with one of int, str, float')

class ArrType():
    size: ty.Optional[int] = None
    ref_type: ty.Type[ty.Union[int, str, float]] = int

class Arr(metaclass=ArrMeta):
    pass

class ArrInt:
    def append(self, value):
        pass

a = Arr[int]()
a.append(3)  # Instance of 'ArrType' has no 'append' member (no-member)

a_t = Arr[int, 5]
print(a_t.size) # Ok
```

### pylint --version output
__main__.py 2.2.2
astroid 2.1.0
Python 3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 22:20:52) [MSC v.1916 32 bit (Intel)]

**Comment by Levitanus** [comment]
Actually, singledispatch worked.
```Python
import typing as ty
from functools import singledispatch


class ArrMeta(type):
    """Var getitem helper metaclass."""

    @singledispatch
    def __getitem__(cls, ref: ty.Type[int]) -> ty.Type['ArrInt']:
        """Return concrete typed class of Arr."""
        if issubclass(ref, int):
            return ArrInt
        if issubclass(ref, float):
            return NotImplemented
        if issubclass(ref, str):
            return NotImplemented
        raise TypeError('subscribe with one of int, str, float')

    @__getitem__.register(tuple)  # type: ignore
    def _(cls, ref: ty.Tuple[ty.Type[int], int]) -> ty.Type['ArrType']:
        if isinstance(ref, tuple):
            size = ref[1]
            _ref = ref[0]
            if size is not None:
                ArrType.size = size
                ArrType.ref_type = _ref
                return ArrType
        raise RuntimeError('There is nothing interesting here')


class ArrType():
    size: ty.Optional[int] = None
    ref_type: ty.Type[ty.Union[int, str, float]] = int


class Arr(metaclass=ArrMeta):
    pass


class ArrInt:
    def append(self, value):
        pass


a = Arr[int]()
a.append(3)  # Instance of 'ArrType' has no 'append' member (no-member)
```

**Comment by Levitanus** [comment]
Thanks for the response. I feel, here is the problem not with annotations, but with the wrong behavior of passing arguments to __getitem__ of metaclass
I was confused by the behavior on these lines:
```Python
if isinstance(ref, tuple):
    size = ref[1]
    ref = ref[0]
    if size is not None:
        ArrType.size = size
        ArrType.ref_type = ref
        return ArrType
```
Anyway, the way to solve the problem is found, and it's not critical :)
